### ISSUE TITLE
Transformer training loss spike/reset with L4 runtime

### ISSUE THREAD

**Issue by bairesearch** [googlecolab/colabtools#5560]
**Describe the current behavior**
One or more transformer training loss spikes/resets are regularly observed with L4 runtime (90% of training runs). The issue occurs approximately every 100000 training samples (apparently random).

**Describe the expected behavior**
No training loss spikes/resets are observed during training (emulating A100 runtime performance).

**What web browser you are using**
Firefox (irrelevant).

**Additional context**
A standard huggingface transformers library is used (custom transformer model), along with the wikipedia dataset. I have confirmed that the issue is not contingent on the dataloader/shuffling, or the precise transformer implementation (encoder/decoder). I have confirmed that the issue is not observed with A100 runtime (on both Colab and locally). 

The attached screenshots show batchIndex on x axis (with batchSize between 32 and 64), and top-1 accuracy on y-axis.

<img width="605" height="340" alt="Image" src="https://github.com/user-attachments/assets/468e389e-22cc-4349-87fe-929902be8734" />
<img width="605" height="340" alt="Image" src="https://github.com/user-attachments/assets/beff701b-9c4b-4074-8325-e5034e24d905" />
<img width="605" height="340" alt="Image" src="https://github.com/user-attachments/assets/65cd75d9-7d3c-41af-81ef-d8e63249557d" />

```

# ---------------------------------------------------------------------------
# 1.  Encoder/Decoder - d_model projection (dense)
# ---------------------------------------------------------------------------

class DenseSnapshotEncoder(nn.Module):
	"""Token DenseSnapshotEncoder -> expected embeddings."""
	def __init__(self, d_input: int, d_model: int, padding_idx: int = 0):
		super().__init__()
		self.emb = nn.Embedding(d_input, d_model, padding_idx=padding_idx)

	def forward(self, tokenProbs: torch.Tensor) -> torch.Tensor:
		return tokenProbs @ self.emb.weight  # (B, L, d_model)

class DenseSnapshotDecoder(nn.Module):
	def __init__(self, embedding: nn.Embedding, padding_idx: int = 0):
		super().__init__()
		self.emb = embedding
		self.padding_idx = padding_idx
		self.bias = nn.Parameter(torch.zeros(embedding.num_embeddings))

	def forward(self, x: torch.Tensor) -> torch.Tensor:
		logits = x @ self.emb.weight.T + self.bias
		return logits

# ---------------------------------------------------------------------------
# 2.  Causal Backbones
# ---------------------------------------------------------------------------

class TransformerEncoderBackbone(nn.Module):

	def __init__(self, d_in: int, max_L: int, nhead: int = 8, nlayers: int = 6, causal: bool = True):
		super().__init__()
		self.pos_emb = nn.Embedding(max_L, d_in)
		d_ff = 4 * d_in
		enc_layer = nn.TransformerEncoderLayer(d_model=d_in, nhead=nhead, dim_feedforward=d_ff, batch_first=True, norm_first=True)
		self.enc = nn.TransformerEncoder(enc_layer, nlayers)
		self.causal = causal
		self._mask_cache = {}	# cache for causal masks keyed by (S, device, dtype)

	def forward(self, x: torch.Tensor, padMask: torch.Tensor) -> torch.Tensor:
		# x: (B, L, d)
		pos_ids = torch.arange(RL2, device=x.device).unsqueeze(0)
		x = x + self.pos_emb(pos_ids)
		if self.causal:
			attn_mask = self._get_causal_mask(RL2, x.device, x.dtype)
			out = self.enc(x, mask=attn_mask, src_key_padding_mask=padMask, is_causal=True)
		else:
			out = self.enc(x, src_key_padding_mask=padMask)
		return out

	def _get_causal_mask(self, S: int, device: torch.device, dtype: torch.dtype) -> torch.Tensor:
		key = (S, device, dtype)
		m = self._mask_cache.get(key)
		if m is None:
			m = torch.triu(torch.ones(S, S, dtype=torch.bool, device=device), 1)
			self._mask_cache[key] = m
		return m
		
# ---------------------------------------------------------------------------
# 3.  Top-level model
# ---------------------------------------------------------------------------

class DenseSnapshotModel(nn.Module):
	def __init__(
		self,
		d_input: int,
		d_target: int,
		d_model: int,
		max_L: int,
		backbone: str,
	):
		super().__init__()
		self.d_model = d_model
		self.encoder = DenseSnapshotEncoder(d_input, d_model, NLPpadTokenID)
		self.backbone = TransformerEncoderBackbone(d_model, max_L)
		self.decoder = DenseSnapshotDecoder(self.encoder.emb, NLPpadTokenID)
		
		def forward(self, x: torch.Tensor, padMask: torch.Tensor, trainOrTest: bool = True) -> Tuple[torch.Tensor, torch.Tensor]:
			x = self.encoder(x)
			enc = self.backbone(x, padMask)
			logits = self.decoder(enc)
			return logits

# ---------------------------------------------------------------------------
# 5. evaluation utilities
# ---------------------------------------------------------------------------

def loss_function(logits: torch.Tensor, targets: torch.Tensor, padMask: torch.Tensor, onehotTargets: bool):
	"""Cross-entropy loss over flattened B1*L dimension with PAD masked out."""
	targets_idx = targets			  # (B, L)
	logits = logits.reshape(-1, logits.size(-1))			 # (B*L, C)
	targets_idx = targets_idx.reshape(-1)					# (B*L)
	loss = F.cross_entropy(logits, targets_idx, ignore_index=NLPpadTokenID)
	return loss
	
def calculate_matches(logits: torch.Tensor, targets: torch.Tensor, onehotTargets: bool) -> float:
	"""Compute bool top-1 accuracy (1/0) for each sample in mini-batch."""
	preds = logits.argmax(dim=-1)	#compare a single target
	valid_mask = (targets != NLPpadTokenID)	#redundant (performed by calculateAccuracy)
	matches = (preds == targets) & valid_mask
	return matches
```

**Comment by metrizable** [comment]
@bairesearch Thanks for reporting the issue and thanks for using Colab.

Thanks for the sample code. Could you provide a minimum reproducible  notebook for easy replication?

**Comment by bairesearch** [comment]
Thank you for your support, but I have confirmed that the issue is not exclusive to the L4 runtime (it can occur with A100 runtime also). There is likely to be a standard pytorch solution to an unidentified bug in the implementation, which I will investigate.

For reference, the attached minimal example appears to consistently fail on L4; when tested twice it failed at a) both batchIndex=2200 and batchIndex=12000, and b) batchIndex=3600 (batchSize=64). It is based on an original codebase of the same configuration that when tested twice also consistently failed on L4; at batchIndex=9200 in both cases (a and b) (batchSize=64). The runtime was disconnected and deleted between all training sessions.

[L4transformerTrainingError.ipynb](https://github.com/user-attachments/files/22048872/L4transformerTrainingError.ipynb)
